pkg/agent/openflow/pipeline.go

```diff
@@ -1982,15 +1987,30 @@ func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAf
 		resubmitTableID = endpointDNATTable
 		lbResultMark = marksRegServiceSelected
 	}
+	endpointsNum := len(endpoints)
+
+	group := c.bridge.CreateGroup(groupID).ResetBuckets()
```

Comment:
Thanks @antoninbas for taking a look at my code. I have to use a fakeGroupId. If I use the same groupID here, the struct of Buckets list in group will be the same address. We can see that in function `func (b *OFBridge) CreateGroup(id GroupIDType)`. For extra endpoints that exceeded the *insertBucketNum* we set(in the code, I set it as 3 for testing), I use a faikeGroupId to "deceive" the function `func (b *OFBridge) CreateGroup(id GroupIDType)` to return with a different group struct. These extra groups is for insert-bucket operation.

```go
func (b *OFBridge) CreateGroup(id GroupIDType) Group {
	ofctrlGroup, err := b.ofSwitch.NewGroup(uint32(id), ofctrl.GroupSelect)
	if err != nil {
		ofctrlGroup = b.ofSwitch.GetGroup(uint32(id))
	}
	g := &ofGroup{bridge: b, ofctrl: ofctrlGroup}
	return g
}

```

```
// Create a new group. return an error if it already exists
func (self *OFSwitch) NewGroup(groupId uint32, groupType GroupType) (*Group, error) {
	// check if the table already exists
	if self.groupDb[groupId] != nil {
		return nil, errors.New("Group already exists")
	}

	// Create a new table
	group := newGroup(groupId, groupType, self)
	// Save it in the DB
	self.groupDb[groupId] = group

	return group, nil
}
```

Reply:
But what about the call to `DeleteGroup`? This one doesn't seem right as it will actually try to delete a group on the switch.
We need a better interface to interact with the bridge here, in order to be able to insert and remove buckets for a group incrementally.

Reply:
The call to `DeleteGroup` will not really delete a group on the switch.  There is a map to record the allocated groupIDs.
The call to `DeleteGroup` only deletes the record of the fakeGroupID from all allocated groupIDs.  

Anyway, if we need insert/remove buckets later, we still need related interface to interact with the bridge.

Reply:
This would be a little obscure to understand and might be error-prone. 
Given all buckets cannot fit single `Add` command, `ofctrl.Group` needs to be changed in some way: either it uses bundle automatically and make clients agnostic, or buckets should be moved to client side, i.e. `openflow.ofGroup` and clients take care of generating bundle messages. In either case, there will still be single `ofctrl.Group` for a group.